geo-types/src/polygon.rs

```diff
@@ -549,3 +549,25 @@ impl<T: AbsDiffEq<Epsilon = T> + CoordNum> AbsDiffEq for Polygon<T> {
         zipper.all(|(lhs, rhs)| lhs.abs_diff_eq(rhs, epsilon))
     }
 }
+
+#[cfg(feature = "rstar")]
+impl<T> ::rstar::RTreeObject for Polygon<T>
+where
+    T: ::num_traits::Float + ::rstar::RTreeNum,
+{
+    type Envelope = ::rstar::AABB<Point<T>>;
+
+    fn envelope(&self) -> Self::Envelope {
+        self.exterior.envelope()
+    }
+}
+
+#[cfg(feature = "rstar")]
+impl<T> ::rstar::PointDistance for Polygon<T>
+where
+    T: ::num_traits::Float + ::rstar::RTreeNum,
+{
+    fn distance_2(&self, point: &Point<T>) -> T {
+        self.exterior.distance_2(point)
```

Comment:
I'm not sure how rstar uses these methods, but this seems incorrect.

A little perusing of the rstar codebase uncovered this usage:
```
    /// Returns `true` if a point is contained within this object.
    ///
    /// By default, any point returning a `distance_2` less than or equal to zero is considered to be
    /// contained within `self`. Changing this default behavior is advised if calculating the squared distance
    /// is more computationally expensive than a point containment check.
    fn contains_point(&self, point: &<Self::Envelope as Envelope>::Point) -> bool {
        self.distance_2(point) <= num_traits::zero()
    }
```

It seems surprising that a point within a polygon would return a >0 distance with the implementation in this PR.

It seems like it'd be nice to use the `EuclideanDistance` trait... but it's in geo.

Reply:
Urgh thanks for catching this @michaelkirk. Time for a EuclideanDistance in private_utils, AKA the module that dare not speak its name?

Reply:
Doing this correctly involves duplicating:

- `Point`-`Polygon` euclidean distance
    - The `CoordinatePosition` logic for `Polygon,` and `CoordPos`

into `private_utils`, which is a **lot** of code. There has to be a better way of doing this.

Reply:
One alternative would be to move the integration onto rstar - adding a a new optional `geo` dependency.

Pros: 
- keeps geo-types small

Cons:
- one time breaking change for the geo-types rstar users.
- maybe annoying to update rstar with latest geo all the time since I think geo breaks more than rstar (but not really that often)

Reply:
How about only implementing `RTreeObject` and not `PointDistance` ?   I guess the nearest-neighbor queries won't be available on `RTree<Polygon>` but that seems fair as it feels quite expensive to do the point-in-polygon computation for this query. 

We could provide the full impl. for simpler cases like `Rect`, so that users can get back nearest-neighbor functionality on polygon via `RTree<GeomWithData<Rect, Polygon>>`  and the `RTree::nearest_neighbor_iter` along with additional logic to then invoke geo algos to find the actual nearest neighbor.

Reply:
Instead of adding to `private_utils`, another option is to create a small `euclidean_distance` crate that `geo` depends on and that `geo-types` conditionally depends on if the `rstar` feature is enabled.

Reply:
That seems like a good compromise that doesn't break anything for anyone. Good idea @rmanoka.

Reply:
Those are both sensible ideas. In terms of long-term maintenance, I'm somewhat in favour of @frewsxcv's (to be clear I'm happy to start off with @rmanoka's, though)

Reply:
Yeah what @frewsxcv suggests is the ideal way for handling private_utils in general, and the current issue.  Just to be clear, that new crate cannot use types from `geo-types` right?  Currently `private_utils` freely uses a lot of geo-types.

Reply:
> Just to be clear, that new crate cannot use types from geo-types right? Currently private_utils freely uses a lot of geo-types.

Hmm true. Euclidean distance also depends on a lot of _other_ algorithms (`Contains`, `Intersects`) and we can't import those without introducing a circular dependency, right?